### PR TITLE
[GLM-4.6V] Support Pipeline Parallelism for GLM-4.6V & GLM-4.1V

### PR DESCRIPTION
**Pull request by yuan-luo**
## Motivation

This PR is to support GLM-4.6V and GLM-4.1V Pipeline Parallelism.
In main branch GLM-4.6V and GLM-4.1V PP don't work correctly. There are two reasons:

1. In pp-size > 1, some pipeline shard doesn't have lm_head weight, but the current code attempt to load it in all shards, so it gets assert with errorkey lm_head.weights. There are some other layers have the similar issue, this PR fixes them. 
Why pp=1 have no this problem, is because GLM-4.1V's tie_word_embeddings=False, then 
a. if pp=1, lm_head is always set to ParallelLMHead()
b. if pp>1, lm_head is only set on the last_rank, for intermediate stage, no lm_head is set.
```
if self.pp_group.is_last_rank:
    if self.pp_group.world_size == 1 and self.config.tie_word_embeddings:
        self.lm_head = self.model.embed_tokens.  
    else:
        self.lm_head = ParallelLMHead(...) <<<<<< last rank enters this branch.
else:
    self.lm_head = PPMissingLayer(). <<<<<< intermediate stage enter this branch, no lm_head
```
Some issues are related with this topic https://github.com/sgl-project/sglang/pull/1508 https://github.com/sgl-project/sglang/issues/2935.
2. general_mm_embed_routine is missing PPProxyTensors parameters.

For the first point, when training/saving checkpoints, some frameworks only save model.embed_tokens.weight and don’t save lm_head.weight separately. 

When the inference framework implements the model structure, some model classes, such as Gemma3_mm, Gemma3n_mm, Granite or some EAGLE draft model, actually define a self.lm_head = nn.Linear(...), and then in tie_weights() make lm_head.weight and embed_tokens.weight point to the same Parameter. There can be cases that even don’t define an lm_head module at all, and instead directly use F.linear(hidden, embed_tokens.weight) in forward to produce the output.

In the current setup, suppose we have pp_size = 2 with GLM-4.1V (where tie_word_embeddings = False is used):
- On the previous stage: There is only PPMissingLayer(), and no lm_head.weight.
- On the last stage: There is a real lm_head module, so lm_head.weight must be loaded.

The modification is when lm_head.weight is encountered there, we choose to skip it, based on the fact that the current rank is not the last rank. 

GLM-4.6V Server:
```
$SGLANG_USE_CUDA_IPC_TRANSPORT=1 python -m sglang.launch_server --model-path /home/admin/GLM-4.6V --mm-attention-backend fa3 --port 30000 --chunked-prefill-size 8192 --disable-radix-cache --disable-overlap-schedule --attention-backend fa3 --tp 4 --pp-size 2
```

Client:
```
$bash bench_n_image.sh 
{"id":"07cdbe828ef6470db12e99d17f63f64a","object":"chat.completion","created":1765271843,"model":"auto","choices":[{"index":0,"message":{"role":"assistant","content":"图中的植物是**刺儿菜**（学名：*Cirsium setosum*），属于菊科蓟属。其典型特征为：茎直立，具纵棱；叶片边缘具刺状齿，两面被蛛丝状毛；头状花序单生于茎顶或枝端，总苞片多层，具刺。  \n\n刺儿菜为多年生草本，常见于田间、路旁、荒地等处，适应性较强。","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":"stop","matched_stop":151336}],"usage":{"prompt_tokens":960,"total_tokens":1058,"completion_tokens":98,"prompt_tokens_details":null,"reasoning_tokens":0},"metadata":{"weight_version":"default"}}
real    0m3.419s
user    0m0.002s
sys     0m0.003s
{"id":"08d083ae213c43478342e73ea6fad238","object":"chat.completion","created":1765271844,"model":"auto","choices":[{"index":0,"message":{"role":"assistant","content":"图中的植物是**刺儿菜**（学名：*Cirsium setosum*），属于菊科蓟属。其典型特征为：茎直立，具纵棱；叶片边缘具刺状齿，两面被蛛丝状毛；头状花序单生于茎顶或枝端，总苞片多层，具刺。  \n\n刺儿菜为多年生草本，常见于田间、路旁、荒地等处，适应性较强。","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":"stop","matched_stop":151336}],"usage":{"prompt_tokens":960,"total_tokens":1058,"completion_tokens":98,"prompt_tokens_details":null,"reasoning_tokens":0},"metadata":{"weight_version":"default"}}
real    0m1.254s
user    0m0.002s
sys     0m0.003s
```

GLM-4.1V
Server:
```
$SGLANG_USE_CUDA_IPC_TRANSPORT=1 python -m sglang.launch_server --model-path /home/admin/GLM-4.1V-9B-Thinking --mm-attention-backend fa3 --port 30000 --chunked-prefill-size 8192 --disable-radix-cache --disable-overlap-schedule --attention-backend fa3 --tp 1 --pp-size 2
```

```
$bash bench_n_image.sh 
{"id":"7bb67e024d4c4230a4891038318b233b","object":"chat.completion","created":1765272124,"model":"auto","choices":[{"index":0,"message":{"role":"assistant","content":"用户现在需要识别图中的植物，从图片看，这种植物有带刺的叶片，可能是刺儿菜（Cirsium setosum），不过也可能是其他蓟属植物。先看特征：叶片边缘有刺，茎和叶有刺，属于菊科蓟属。刺儿菜学名Cirsium setosum（有时也叫Cirsium arvense？不对，要确认）。不过更常见的可能是刺儿菜，学名Cirsium setosum（Beauv.）Bess.，别名小蓟、刺菜等。需要确认形态：叶片羽状深裂，边缘有刺，茎直立，有刺，符合蓟属特征。所以判断为刺儿菜，学名Cirsium setosum。图中的植物是刺儿菜，其学名是 **Cirsium setosum**（也常被称为 *Cirsium arvense* ，不过二者在分类和形态上有一定差异，需结合地域等判断，刺儿菜更偏向 *Cirsium setosum* 这类形态，叶片等特征符合刺儿菜常见形态）。刺儿菜属于菊科蓟属植物，常生长在草地、路边等处，叶片边缘和茎部有刺，是常见的野生草本植物，部分地区也用于药用等。","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":"stop","matched_stop":151348}],"usage":{"prompt_tokens":956,"total_tokens":1237,"completion_tokens":281,"prompt_tokens_details":null,"reasoning_tokens":0},"metadata":{"weight_version":"default"}}
real    0m3.268s
user    0m0.003s
sys     0m0.001s
{"id":"0379a48690704a5ab774dd4d6e6ed131","object":"chat.completion","created":1765272127,"model":"auto","choices":[{"index":0,"message":{"role":"assistant","content":"用户现在需要识别图中的植物，从图片看，这种植物有带刺的叶片，可能是刺儿菜（Cirsium setosum），不过也可能是其他蓟属植物。先看特征：叶片边缘有刺，茎和叶有刺，属于菊科蓟属。刺儿菜学名Cirsium setosum（有时也叫Cirsium arvense？不对，要确认）。不过更常见的可能是刺儿菜，学名Cirsium setosum（Beauv.）Bess.，别名小蓟、刺菜等。需要确认形态：叶片羽状深裂，边缘有刺，茎直立，有刺，符合蓟属特征。所以判断为刺儿菜，学名Cirsium setosum。图中的植物是刺儿菜，其学名是 **Cirsium setosum**（也常被称为 *Cirsium arvense* ，不过二者在分类和形态上有一定差异，需结合地域等判断，刺儿菜更偏向 *Cirsium setosum* 这类形态，叶片等特征符合刺儿菜常见形态）。刺儿菜属于菊科蓟属植物，常生长在草地、路边等处，叶片边缘和茎部有刺，是常见的野生草本植物，部分地区也用于药用等。","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":"stop","matched_stop":151348}],"usage":{"prompt_tokens":956,"total_tokens":1237,"completion_tokens":281,"prompt_tokens_details":null,"reasoning_tokens":0},"metadata":{"weight_version":"default"}}
real    0m3.097s
user    0m0.001s
sys     0m0.003s
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: test/srt/test_pp_single_node.py [hunk 1]
```diff
@@ -318,5 +319,42 @@ def test_chunked_prefill_with_small_bs(self):
         )
 
 
+class TestGLM41VPPAccuracy(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.model = DEFAULT_MODEL_NAME_FOR_TEST_GLM_41V_PP
+        cls.base_url = "http://127.0.0.1:23333"
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply test/srt/test_pp_single_node.py]: Addressed.

File: python/sglang/srt/models/glm4v.py [hunk 2]
```diff
@@ -766,6 +770,10 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
                 # Skip loading extra bias for GPTQ models.
                 if name.endswith(".bias") and name not in params_dict:
                     continue
+
+                if name not in params_dict:
```
Comment (BBuf) [inline python/sglang/srt/models/glm4v.py:785]: Please add a comment—it's not easy to understand.

File: python/sglang/srt/models/glm4v.py [hunk 3]
```diff
@@ -779,6 +787,10 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
                     # Skip loading extra bias for GPTQ models.
                     if name.endswith(".bias") and name not in params_dict:
                         continue
+
+                    if name not in params_dict:
```
Comment (BBuf) [inline python/sglang/srt/models/glm4v.py:802]: Same as above.

File: python/sglang/srt/models/glm4v.py (hunk 2 above)
Comment (BBuf) [reply python/sglang/srt/models/glm4v.py:785]: Generated by gemini-3.0-pro, use it as ref:

```python
if name not in params_dict:
      if self.pp_group.world_size > 1:
          skipped_weights.append(name)
          logger.debug(
              f"Rank {self.pp_group.rank_in_group}/{self.pp_group.world_size}: "
              f"Skipping weight '{name}' (not present in this pipeline stage)"
          )
          continue
      else:
          raise KeyError(f"Parameter {name} not found in model")
```

Comment (yuan-luo) [reply python/sglang/srt/models/glm4v.py:785]: Comments added in the head of the loop, covering both parts.

File: python/sglang/srt/models/glm4v.py (hunk 3 above)
Comment (yuan-luo) [reply python/sglang/srt/models/glm4v.py:802]: reverse ditto

**Review by BBuf — APPROVED** [review]
LGTM.

File: python/sglang/srt/models/glm4v.py [hunk 4]
```diff
@@ -750,13 +752,26 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
...
+        # If this name does not exist in the current rank’s params_dict,
+        # it does not belong to this pipeline stage, thus we simply continue.
         for name, loaded_weight in weights:
             if "rotary_emb.inv_freq" in name:
                 continue
             if "language_model" in name:
                 name = name.replace(r"model.language_model.", r"model.")
             if "model.visual." in name:
                 name = name.replace("model.visual.", "visual.")
+            if name.startswith("lm_head.") and not self.pp_group.is_last_rank:
```
Comment (byjiang1996) [inline python/sglang/srt/models/glm4v.py:773]: Curious. Just to double check. Are the load_weights() changes required in glm4v_moe.py too? GLM4v have both dense (hosted in glm4v.py) and MoE (hosted in glm4v_moe.py) architecture

Comment (yuan-luo) [reply python/sglang/srt/models/glm4v.py:773]: Similar change might be required in GLM4V_MoE as well. But I haven't verified it yet. Will follow up it in a new PR.

**Review by byjiang1996 — APPROVED** [review]
